### ISSUE TITLE
Expressing a contractual promise to accept null (categorically or conditionally): a requirement for 1.0?

### ISSUE THREAD

**Issue by kevinb9n** [jspecify/jspecify#113]
This is intended to be a new framing and starting point for the same issue recently discussed under the headings "the specification for nullable", "semantics of nullable", "specifying predicate nullness", etc. Like other issues, it can move into a document IF that proves necessary.

I am hopeful that everyone can accept this particular framing of the issue -- or give a clear argument for why a different framing will help us move forward better.

If the framing appears too much to predestine a particular decision, that *could* happen because it's a good decision, but it might also just mean we need to add the next layer onto the model ("Now, here's where this breaks down"). But, crossing fingers that we won't just throw it out. Enough said by now.

THIS IS ABOUT 1.0

I think it is best to frame this as a requirements question *for 1.0*, specifically. The outcome I think we seek is a decision between listing this feature area as an explicit requirement, or listing it as an explicit non-requirement-for-now.

Please note that discussion of how we would design the feature is *usually* not relevant to such a decision -- but, in cases, elements of that discussion *can* come into play.

A "yes, make it a 1.0 requirement" answer means more work for us, but I can think of three [four] basic reasons for deciding "yes":

1. We judge the combined risk to be high that we will both (a) want to add the feature post-1.0, and (b) find that our 1.0 design decisions have prevented us from doing so.

2. We judge that too many libraries may not see the product as even worth adopting without the feature, and our 1.0 release will fizzle as a result.

3. Multiple partners believe they will feel compelled to proceed independently on this, in a way that could cause appreciable and harmful fracturing, yet it would be unreasonable to do it as a separate complementary project.

4. [Edit, thanks @lazaroclapp] Even one partner is forced to conclude they can't use our annotations profitably at all (not even by way of `@Implies` (#35)), or that what their users need to do is too painful.

WHAT WE'RE NOT DEBATING

The basic task I think we're all signed up for is to separate "nullable Foo" from "non-nullable Foo" *as types*, allowing appearances of the Foo type in APIs to signify their null-inclusion intentions. (Let's sidestep "unspecified" as a mostly orthogonal distraction here.)

The most important effect of this type separation is that tools can treat dereferencing as if it is an operation belonging to the non-nullable type only.

By analogy, we might imagine that we can create a new subtype of `String` called `NonEmptyString`, and arrange that all values of the `String` type except for `""` will always have this new type. Some operations, like `char firstChar()`, would belong only to the new subtype. Methods can now choose to require the subtype to be passed in. A method can promise _never_ to return `""` simply by changing its return type to `NonEmptyString` (even if its supermethod doesn't).

For any community plagued with `EmptyStringException` stack traces, it doesn't seem a tall claim to say that this much would already make a big difference. That exception wouldn't disappear entirely, but we wouldn't expect it to, because types only do so much.

The part of our work that is (I hope) non-controversial is just this type bifurcation (really, trifurcation) part.

WHAT (I THINK) WE ARE DEBATING

Methods have contracts with many more nuances than their type signatures alone can convey. For example, using the NonEmptyString analogy, a method must accept `String` so long as there exists *any* circumstance in which `""` is valid. But it has never (officially) had a way to express in its contract what these circumstances are, or even the basic fact of whether it pledges to _definitely_ accept that value. It's only if it will _never_ accept the value that adopting `NonEmptyString` gives it a path forward.

CAVEAT ABOUT THE ANALOGY

Analogies can always be seen as trivializing. In real life, there is no big problem with `EmptyStringException` but there _is_ a huge problem with `NullPointerException`. The analogy is not meant to suggest a lower importance of solving these use cases.

BACKGROUND

This issue is attempting to serve as the new starting point for the questions raised here:
[First thread](https://groups.google.com/g/jspecify-dev/c/xY2QpnyNW8Q/m/yt3jsNKeAgAJ)
[Werner's doc](https://docs.google.com/document/d/1FSRi7KdyR42hTt3yDx9-OUGk9w1vH0fDSG6YCem7QV8/edit)
[Lazaro's doc](https://docs.google.com/document/d/1EIToKowyghVMXCaf_AzLCKsQ_nN4EZW4x-dmBXkycT4/edit)

**Comment by kevinb9n** [comment]
I am claiming that it is fair to treat this as a question of whether to _add_ a new requirement to the project. As such, it has a clear [base option](https://docs.google.com/document/d/1tLIjEuNhABtGp7WSy1_wFeieaI4DjynVJ8QJ95lYfJ0/edit#bookmark=id.y9dp2149s361) -- unless argued successfully otherwise, it is by default not in scope. 

I think that the the three criteria above are an appropriate map for crafting an argument to add the requirement. Does anyone want to put forth and champion that argument?

If not, we should state it as a non-requirement and close this issue for now.

**Comment by lazaroclapp** [comment]
tl;dr: I am fine with this not being a requirement for 1.0 iff this is not a requirement for CF to use jspecify annotations in a way consistent with other tools.

SIDE NOTE ON THE ANALOGY

I still feel the analogy doesn't resolve the question, because both the nullness type system and the string-emptiness type system are clear when _all_ code is checked against those type systems, and have similar issues when some methods are expected to use those types for their signature but _not_ have their implementation type-checked. In theory you could say that a non-type-checked method and one that does an explicit `(NonEmptyString) s` cast are equivalent, but I still feel that there is a difference. For a proper analogy, the difficulties of the docs above would arise when _retrofitting_ String/NonEmptyString into the JDK API signatures (e.g. via a jdk.stub like the one used by CF), just the same as when adding nullability info to those same APIs without changing their code.

BUT... that's besides the point. I think we agree in the choices we have, and the requirements in tension, I am just not sure the analogy really argues for one decision or the other.

PROPOSED DECISION CRITERIA

I think the fundamental criteria I would like to take this choice based on is this: Can jspecify 1.0 be reasonably used by sound/sounder tools like the Checker Framework, perhaps by _augmenting_ its annotations with special case `@Contract`/`@Precondition`/`@Postcondition` annotations that are _seldom_ needed? 

If such tool would need a `@ReallyNullableIMeanIt` annotation to go along with `@Nullable` even 30% of the time, then that's a big red flag for jspecify's generality for me.

MY TAKE

My most recent discussions with Werner seem to indicate that the above worst case is not true, and that CF can make some _reasonable_ use of jspecify 1.0 with binary+unspecified nullability. I'd let him and Michael pitch in, but if they don't have a strong objection, neither do I. 

Just to clarify, if I am solely acting on behalf of the NullAway project and/or Uber, then we are perfectly happy to use the existing `@Nullable/@NonNull` semantics, at least in regards to the issue under discussion. I just want the Checker Framework folks to also be happy with the outcome here, since that improves the chances of jspecify becoming "The Standard" (tm)

All this with the caveat that post-1.0 it does seem like a good idea to have some mechanism to specify conditional nullability, since both CF and NullAway, as well as (I believe) IntelliJ, arrived to something related in their design at some point :)

**Comment by kevinb9n** [comment]
This may be subtle, but I really am trying not to use the analogy as if it is an *argument* in and of itself. Just a mental model for what it means to perform the kind of type bifurcation we are doing. And you're right that it does not account for the issue of unchecked code, and in fact my own thought processes are still highly undeveloped (and so you've had to try to bring that perspective numerous times). I plan to follow up separately about that.

Next, thanks very much for reminding us that even if we don't _have_ this 1.0 requirement, we still have a requirement to _support_ a product that does take on the requirement. It is hard to make what that means very concrete, but it must feel like that product is able to build on to what we specify rather than having to cast it aside. That is important, I believe.

**Comment by kevinb9n** [comment]
Lazaro, I've just filed #115 to try to get some clarification on how we should treat issues unique to unchecked code.

**Comment by kevinb9n** [comment]
I think we can consider this decided (NOT a requirement) unless/until new information comes to light.
